content/post/2023/14-solving-a-sudoku-with-symbiyosys-and-formal-verification.md

```diff
@@ -0,0 +1,305 @@
+assume(sudoku_grid[7][6]==6);
+//line 9
+assume(sudoku_grid[8][1]==4);
+assume(sudoku_grid[8][5]==9);
+assume(sudoku_grid[8][6]==3);
+assume(sudoku_grid[8][8]==8);
+end
+```
+
+### Hijacking the use of the SymbiYosys
+
+With the game rules handed over to our solver, we want it to return the solution now. Here lies our ingenious twist in using the tool!
+
+As highlighted before, in a correctly solved Sudoku, every row, column, or box's digit sum is 45. If we were to declare: "assert(sum_line=6'd45);" the solver would merely confirm with a "yes"... and leave it at that.
+
+Yet, our goal is the solution. So, we'll assert the contrary. We'll claim the sum of the first row's digits can't possibly be 45. In response, the solver produces a counter-proof where the first row's digits indeed sum up to 45. Given it has to respect the Sudoku rules, this counter-proof elegantly unfolds as our desired Sudoku solution.
+```verilog
+// Assert that sum cannot be equal to 45
+wire [5:0] sum;
+assign sum = sudoku_grid[0][0] +
+			 sudoku_grid[0][1] +
+			 sudoku_grid[0][2] +
+			 sudoku_grid[0][3] +
+			 sudoku_grid[0][4] +
+			 sudoku_grid[0][5] +
+			 sudoku_grid[0][6] +
+			 sudoku_grid[0][7] +
+			 sudoku_grid[0][8];
+
+always @(*) begin
+	assert(sum!=6'd45);
+end
+`endif FORMAL
+endmodule
+```
+
+## Let's run the verification
+
+To install SymbiYosys and all the required formal solvers, I recommend following the straightforward installation process provided by [oss-cad-suite](https://github.com/YosysHQ/oss-cad-suite-build).
+
+To execute the verification, we'll need our sudoku.v file and a SymbiYosys configuration file named sudoku.sby with the content below:
+```
+[options]
+mode prove
+depth 20
+
+[engines]
+smtbmc
+
+[script]
+read -formal sudoku.v
+prep -top sudoku
+
+[files]
+sudoku.v
+```
+
+All we have to do is run the following command :
+```
+sby -f sudoku.sby
+```
+
+And as you can see... it fails... WAIT! WHAT?
+
+![image](/static-2023/assumption_unsatisfiable.png)
+_SymbiYosys log_
+
+"Assumptions are unsatisfiable." What does that mean?
+
+- The line assumptions are correct.
+- The row assumptions are correct.
+- The box assumptions are correct.
+- The... line 2 of the initialization is wrong...
+
+```
+assume(sudoku_grid[1][5]==3); // wrong
+assume(sudoku_grid[1][4]==3); // correct
+```
+
+When I visited my grandfather, I handed him the original and copied it onto a paper to solve it myself... I made a copying error(deep breath)... At least we stumble upon a unexpected feature : we now know how to identify an unfeasible grid !  
+
+Now, having corrected my error... it still fails, but as expected: SymbiYosys notifies me that my assertion didn't hold! And that there's proof! Which is precisely what I was aiming for.
+
+Here's the solution viewed through gtkwave (SymbiYosys give you the path of the vcd file):
+![image](/static-2023/unreadable_solution.png)
+_GTKWave screenshot_
+
+Hmm... that's not very user-friendly to interpret.
```

Comment:
There's work-in-progress to support `$display` and similar during formal verification with SBY. That way you could write a loop in verilog that can directly access the signals to output the solution formatted as a grid from within SBY. It's not quite ready yet, but I think your Sudoku solver would be a really nice example for it and I guess we could do a follow up post as soon as that features is done.